core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java

```diff
@@ -145,6 +154,21 @@ protected Boolean addBooleanResponse(Address sender, Boolean response) {
       }
    }
 
+   @Override
+   public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throws Throwable {
+      if (!isLocalModeForced(command)) {
+         if (command.hasAnyFlag(FlagBitSets.SKIP_SIZE_OPTIMIZATION) || !usingStores()) {
+            return asyncValue(clusterPublisherManager.keyReduction(false, null, null, ctx,
+                  command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
+                  PublisherReducers.add()));
+         }
+
+         return asyncValue(clusterPublisherManager.sizePublisher(command.getSegments(), ctx, command.getFlagsBitSet()));
+      }
+
+      return invokeNext(ctx, command);
```

Comment:
I would expect this to look more like this. The `sizePublisher` would then pass the FLAG_MODE_LOCAL flag.

```suggestion
      if (isLocalModeForced(command) {
         return invokeNext(ctx, command);
      }
      if (command.hasAnyFlag(FlagBitSets.SKIP_SIZE_OPTIMIZATION)) {
         return asyncValue(clusterPublisherManager.keyReduction(false, null, null, ctx,
               command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
               PublisherReducers.add()));
      }

      return asyncValue(clusterPublisherManager.sizePublisher(command.getSegments(), ctx, command.getFlagsBitSet()));
```

Reply:
I am not sure why the usingStores was referenced. We should be able to do the optimization with or without stores.